libraries/AP_Scripting/applets/plane_follow.lua

```diff
@@ -0,0 +1,982 @@
+end
+local function disable_follow_mode()
+   if not follow_enabled then
+      return
+   end
+   -- Follow switched to disabled - return to EXIT mode - but not if disarmed
+   if not arming:is_armed() then
+      vehicle:set_mode(exit_mode)
+   end
+   follow_enabled = false
+   gcs:send_text(MAV_SEVERITY.INFO, SCRIPT_NAME_SHORT .. ": disabled")
+end
+
+--[[
+   follow_check() - check for user activating follow using an RC switch 
+      - set HIGH and if so
+         - switches to GUIDED
+         - enables follow
+         - resets the PID controllers
+      -- set LOW and if so
+         - exits from GUIDED to the FOLLP_EXIT_MODE
+         - disables follow
+--]]
+local last_follow_active_state = rc:get_aux_cached(FOLLP_ACT_FN:get())
+local function follow_check()
+   if FOLLP_ACT_FN == nil then
+      return
+   end
+   local foll_act_fn = FOLLP_ACT_FN:get()
+   if foll_act_fn == nil then
+      return
+   end
+   local active_state = rc:get_aux_cached(foll_act_fn)
+   if (active_state == last_follow_active_state) then
+      return
+   end
+   if( active_state == 0) then
+      disable_follow_mode()
+   elseif (active_state == 2) then
+      enable_follow_mode()
+   end
+   last_follow_active_state = active_state
+end
+
+local function wrap_360(angle)
+   local res = math.fmod(angle, 360.0)
+    if res < 0 then
+        res = res + 360.0
+    end
+    return res
+end
+
+local function wrap_180(angle)
+    local res = wrap_360(angle)
+    if res > 180 then
+       res = res - 360
+    end
+    return res
+end
+
+-- convert a groundspeed to airspeed using windspeed and EAS2TAS (from AHRS)
+local function calculate_airspeed_from_groundspeed(velocity_vector)
+   --[[ 
+   This is the code from AHRS.cpp    
+   Vector3f true_airspeed_vec = nav_vel - wind_vel;
+
+   This is the c++ code from AP_AHRS_DCM.cpp and also from AP_AHRS.cpp
+   float true_airspeed = airspeed_ret * get_EAS2TAS();
+        true_airspeed = constrain_float(true_airspeed,
+                                        gnd_speed - _wind_max,
+                                        gnd_speed + _wind_max);
+        airspeed_ret = true_airspeed / get_EAS2TAS(
+   --]]
+
+   local airspeed_vector = velocity_vector - windspeed_vector
+   local airspeed = airspeed_vector:length()
+   airspeed = airspeed * ahrs_eas2tas
+   airspeed = constrain(airspeed, airspeed - windspeed_max, airspeed + windspeed_max)
+   airspeed = airspeed / ahrs_eas2tas
+
+   return airspeed
+end
+
+-- ChatGPT code to calculate the distance to the target along_track and cross_track
+local function calculate_track_distance(P_f, P_l)
+    -- Get the follower's ground-relative velocity
+    local V_f = ahrs:get_velocity_NED()
+    if V_f == nil or (V_f:x() == 0 and V_f:y() == 0) then
+        -- No valid velocity, return zeros
+        return 0, 0
+    end
+
+    -- Create horizontal velocity vector
+    local D_f = Vector3f()
+    D_f:x(V_f:x())
+    D_f:y(V_f:y())
+    D_f:z(0)
```

Comment:
use Vector2f if Z is zero on both.

Reply:
I need to do a dot product with another Vector3f a few lines down